### ISSUE TITLE
overridden_fields should not lint if more specific type is used

### ISSUE THREAD

**Issue by a14n** [dart-lang/linter#622]
The following code snippet from Flutter has a lint on _FlutterError#message_ but shouldn't IMHO.

```dart
class AssertionError extends Error {
  final Object message;
  AssertionError([this.message]);
  String toString() => "Assertion failed";
}

class FlutterError extends AssertionError {
  FlutterError(this.message);
  @override
  final String message;
  @override
  String toString() => message;
}
```

**Comment by JPaulsen** [comment]
Why not @a14n? The rule says:

**DON'T** override fields.

That means that should lint all overriden fields even if they are useful or not?

**Comment by a14n** [comment]
If I remove the field in the child class I loose the type information that _message_ is a _String_.

BTW I can't find this rule in [Effective Dart](https://www.dartlang.org/guides/language/effective-dart).

**Comment by zoechi** [comment]
That was a DDC and strong-mode restriction that overriding fields was not supported, which was fixed recently.
I'm pretty sure there never was a Dart style guide rule.
It's still discouraged because of weird issues it can cause with memory usage and some code might see only the field in the superclass while other code might see the subclass' field.

**Comment by a14n** [comment]
Thanks for the explanation. I will try to refactor the code and perhaps close this issue.

**Comment by alexeieleusis** [comment]
This lint does not come from effective dart but rather from https://github.com/dart-lang/sdk/issues/25567 I don't think it is intended that rules come from effective dart.

That myself have had cases where I want to tighten the type on the extending class, can it be annotated in the base class? like with `@checked` of `covariant`, I honestly don't know, that's why I ask. Do we want any other signal than tightening the type? How will the constructors be affected in the derived class? Should that particular case be implemented as an abstract getter where the proposed annotations work?

**Comment by Hixie** [comment]
The FlutterError code above is bogus. It's introducing a new field which means the object has two slots but should only have one.

It should just say:
```dart
   @override
   String get message => super.message;
```
...and the constructor should pass `message` up to the superclass.

When we wrote FlutterError, AssertionError didn't have a message field.

**Comment by a14n** [comment]
@Hixie thanks
